CONTRACTS: Documenting semantic of loop invariant synthesizers (#7089)
diffblue/cbmc: doc/cprover-manual/loop-invariant-synthesizer.md

[CPROVER Manual TOC](../)

## Loop Invariant Synthesis with goto-instrument

This is a loop invariant synthesizer built in goto-instrument that 

> synthesize [invariant clause](contracts-invariants.md) for a 
> pre-existing GOTO binary, annotate the invariant clause into 
> the GOTO binary, and produce the annotated binary as output.

Reviewer: ```suggestion
synthesize [loop invariants](contracts-invariants.md) for a 
pre-existing GOTO binary, annotate the invariant into 
the GOTO binary, and produce an annotated binary as output.
```

> Loop invariant synthesized by the synthesizer must be inductive 
> in the sense that 

Reviewer: ```suggestion
All synthesized loop invariants must be inductive, i.e., 
```

1. the instrumented loop invariant checks success, no matter other 
assertions fail or success;
2. the base-case is from the whole GOTO program with a main 
function as the entry point.

> 

Reviewer: Could you provide an example here?

Reviewer 2: I find these two points somewhat unclear. An example might help, as Felipe mentioned, especially for point 1.

I would probably just remove point 2. "`main` function" need not be the entrypoint for CBMC (can be overriden), and the base case is no different from any other symex path -- the path constraint along a path that reaches the loop head _is_ the base case.

It might be worth adding a pointer to semantics of loop invariants ([contracts-invariants.md#semantics](https://github.com/diffblue/cbmc/blob/develop/doc/cprover-manual/contracts-invariants.md#semantics)), which already covers the induction principle.

Author: The first point I want to make here is that we only care about if the invariant is inductive, but not other checks (even if there is `assert(0);` in the loop). The purpose of the second point is to distinguish: 1) the invariants are inductive for any inputs of the function, and 2) the invariants are inductive for only paths in the given GOTO binary.

As an example, ```k == 1``` is an inductive invariant in the following program. However, it is not always inductive if we only consider the function ```foo```---it doesn't hold when the parameter ```k``` is not 1.
```c
void foo(int k)
{
  for(int i = 0; i < 5; i++)
  {
    if(k != 1)
      k = 2;
  }
}

void main()
{
  foo(1);
}
```

You are right that the semantics of loop invariants is enough to illustrate the point. I will update the text and add a link to the semantics of loop invariants.

Reviewer: I still few this section needs an example. Take a small program, check the program with pointer checks enable using a bound, synthesize the invariant, check the problem again without the bound. One of the most important part of the documentation is to walk potential users and future developers in the process of how to use this tool.

Reviewer 3: I like the idea of examples.  It would also be good to discuss what the initial state is as part of defining / saying they are inductive.

Reviewer 4: If the invariants are inferred in the context of a proof harness that does not fully cover the input space of the function that contains the loop, is there a risk that we could assume the invariant in a different context where it could actually not hold without realizing it?
Or would we attempt to re-prove the inferred invariant in each context in which the function that contains the loop gets called ?

Reviewer 3: It could happen but it should not be silent as the check for whether it is a loop invariant should fail in the new context.  I presume we have tests for "`__CPROVER_invariant` that aren't actually true?

> For different synthesis strategy, there are different prerequisites

Reviewer 2: ```suggestion
For different synthesis strategies, there are different restrictions
```

on the input GOTO binary, and different guarantees on the synthesized
loop invariants, summarized in the

> following table. 

Reviewer 2: Why do we have this table? Are there other strategies? If so, why is the flag for "enumerative" strategy called "--synthesize-loop-invariants"? Shouldn't it be called something like `--synthesize-loop-invariants-enumerative"?

Author: There are no other strategies now. But I expect at least we will have a CHC-based synthesizer as another strategy. 

Yes, I agree that the flag should be more specified when there are more than one strategies. However, I think it would be better to keep is simple now.

Reviewer 2: I am not sure if the CHC-based synthesizer also aims to plug the invariant back into GOTO, and whether it would follow the same workflow as the enumerative one.

I agree on the simplicity part, but users might expect `--synthesize-loop-invariants` to run a portfolio of all available synthesizers. To avoid ambiguity and churn in future, I would suggest renaming the flag to be more specific and explicit.

Author: I see. I modified the text and also the name of the flag in code.

> Strategy         | Flag                                           | Prerequisites                                      | Guarantees

Reviewer 2: ```suggestion
Strategy         | Flag                                           | Restrictions                                      | Guarantees
```

-----------------|------------------------------------------------|----------------------------------------------------|----

> Enumerative      | --synthesize-loop-invariants                   | The input GOTO binary contains only pointer checks | All checks success in the annotated GOTO binary

Reviewer: ```suggestion
Enumerative      | --synthesize-loop-invariants                   | The input GOTO binary contains only pointer checks | All checks will succeeded in the annotated GOTO binary
```

> Enumerative      | --synthesize-loop-invariants                   | The input GOTO binary contains only pointer checks | All checks success in the annotated GOTO binary

Reviewer 2: I am unclear on the restrictions part here. When you say "contains only pointer checks," are you excluding all arithmetic checks (overflows, conversions etc.)? Does pointer checks include all kinds of pointer checks (overflows, primitive checks etc.) or just pointer validity checks?

Should they not appear anywhere in the goto file, or just the loop body? What is the expected behavior when they do appear, do we error out?

Author: I refer by _pointer checks_ the checks added by goto-instrument with the flag ```--pointer-check```. I will clarify it.

When there checks are than pointer checks, there are two possibility. 
1. Those checks don't depend on any loop invariant. The synthesizer will work as those checks don't exist. 
2. Those checks depend on some loop invariant. We will warn the user that there exists non-pointer-checks and the enumeration is likely not to terminate.

Reviewer 2: I see, thanks. Let's add this behavior to the document

Author: To be more accurate, I removed the restriction part. Instead we say the synthesizer is _designed_ for some kind of GOTO binaries---it works best for them, but still have a chance to solve problems with other checks.

> ## Additional Resources

Reviewer: Are there any papers or external material about how this synthesis work that you could reference here too?

Author: Thank you for the review! 
There is not a perfect fit here. I link a related paper, and will write something in the training material which we can link later.

[CPROVER Manual TOC](../)

> ## Loop Invariant Synthesis with goto-instrument

Reviewer: ```suggestion
## Loop Invariant Synthesis
```

Reviewer 5: This is marked as resolved but hasn't actually been changed?

This is a loop invariant synthesizer built in goto-instrument that 
synthesize [loop invariants](contracts-invariants.md) for a 
pre-existing GOTO binary, annotate the invariant clause into 

> the GOTO binary, and produce an annotated binary as output.

Reviewer 5: Is it really most important to talk about `goto-instrument` here? I'd expect that users first want to know a sentence about what loop invariants are (and, specifically, what kinds of loop invariants are being synthesized - inductive ones?), and then what capabilities/limitations this synthesizer has (it's an undecidable problem).

All synthesized loop invariants must be inductive, i.e., 
all [loop-invariant checks](contracts-invariants.md#semantics)
succeed in the instrumented GOTO binary.

For different synthesis strategies, there are different guarantees 
beyond the inductiveness on the synthesized loop invariants, 
summarized in the following table.

Strategy         | Flag                                           | Guarantees
-----------------|------------------------------------------------|----

> Enumerative      | --synthesize-loop-invariants-enumerative       | All checks will succeeded in the annotated GOTO binary

Reviewer 3: What happens if the asserts (I presume that is what you mean by checks) are false?

### Enumerative Synthesizer

> This synthesizer is designed for GOTO binary that contains only checks instrumented 
> with the flag  ```--pointer-check```, and synthesizes loop invariants that make
> all checks succeeded. The idea of the synthesizer is first inferring a clause, 
> using the idea of weakest preconditions, for each violated check to make it succeed, 
> and then enumerating a strengthening clause to make the inferred clause inductive.

Reviewer: ```suggestion
The enumerative synthesizer is designed for GOTO binaries instrumented only for pointer checks
(i.e., using ```--pointer-check``` flag), because it synthesizes loop invariants that make
all pointer checks succeeded. First, the synthesizer infer a clause for each violated check
(based on weakest preconditions) in order to make them succeed. Then, it enumerates
a strengthening clause to make it inductive.
```

> If the input GOTO binary contains a violated check that is not a pointer check, 
> the synthesizer will skip the inferring step and simply enumerate predicate that 
> is inductive and can make the check succeed.

Reviewer: ```suggestion
If the input GOTO binary contains a violated checks beyond pointer checks, 
the synthesizer will skip the inferring step and simply enumerate predicates that 
are inductive and can make the checks succeed.
```
